cmd/pebble/cmd_exec.go

```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		group = cmd.Group
+	} else {
+		groupID = &gid
+	}
+
+	// Specify UseTerminal if -t/--terminal is given, or if both stdin and
+	// stdout are TTYs.
+	stdinIsTerminal := ptyutil.IsTerminal(unix.Stdin)
+	stdoutIsTerminal := ptyutil.IsTerminal(unix.Stdout)
+	var useTerminal bool
+	if cmd.Terminal {
+		useTerminal = true
+	} else if cmd.NoTerminal {
+		useTerminal = false
+	} else {
+		useTerminal = stdinIsTerminal && stdoutIsTerminal
+	}
+
+	// Record terminal state (and restore it before we exit).
+	if useTerminal && stdinIsTerminal {
+		oldState, err := ptyutil.MakeRaw(unix.Stdin)
+		if err != nil {
+			return err
+		}
+		defer ptyutil.Restore(unix.Stdin, oldState)
+	}
+
+	// Grab current terminal dimensions.
+	var width, height int
+	if stdoutIsTerminal {
+		width, height, err = ptyutil.GetSize(unix.Stdout)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Start the command.
+	opts := &client.ExecOptions{
+		Command:     command,
+		Environment: env,
+		WorkingDir:  cmd.WorkingDir,
+		Timeout:     cmd.Timeout,
+		User:        user,
+		UserID:      userID,
+		Group:       group,
+		GroupID:     groupID,
+		UseTerminal: useTerminal,
+		Width:       width,
+		Height:      height,
+		Stdin:       os.Stdin,
+		Stdout:      os.Stdout,
+	}
+	if !useTerminal {
+		opts.Stderr = os.Stderr
+	}
+	process, err := cmd.client.Exec(opts)
+	if err != nil {
+		return err
+	}
+
+	// Start a goroutine to handle signals and window resizing.
+	done := make(chan struct{})
+	defer close(done)
+	go execControlHandler(process, useTerminal, done)
+
+	// Wait for the command to finish.
+	err = process.Wait()
+	switch e := err.(type) {
+	case nil:
+		return nil
+	case *client.ExitError:
+		logger.Debugf("Process exited with return code %d", e.ExitCode())
+		return &exitStatus{e.ExitCode()}
+	default:
+		return err
+	}
+}
+
+func execControlHandler(process *client.ExecProcess, useTerminal bool, done <-chan struct{}) {
+	ch := make(chan os.Signal, 10)
+	signal.Notify(ch,
+		unix.SIGWINCH, unix.SIGHUP,
+		unix.SIGTERM, unix.SIGINT, unix.SIGQUIT, unix.SIGABRT,
+		unix.SIGTSTP, unix.SIGTTIN, unix.SIGTTOU, unix.SIGUSR1,
+		unix.SIGUSR2, unix.SIGSEGV, unix.SIGCONT)
+
+	for {
+		var sig os.Signal
+		select {
+		case sig = <-ch:
+		case <-done:
+			return
+		}
+
+		switch sig {
+		case unix.SIGWINCH:
+			if !useTerminal {
+				logger.Debugf("Received SIGWINCH but not in terminal mode, ignoring")
+				break
+			}
+			logger.Debugf("Received '%s signal', updating window geometry", sig)
+			width, height, err := ptyutil.GetSize(unix.Stdout)
+			if err != nil {
+				logger.Debugf("Error getting terminal size: %v", err)
+				break
+			}
+			logger.Debugf("Window size is now: %dx%d", width, height)
+			err = process.SendResize(width, height)
+			if err != nil {
+				logger.Debugf("Error setting terminal size: %v", err)
+				break
+			}
+		case unix.SIGHUP:
+			file, err := os.OpenFile("/dev/tty", os.O_RDONLY|unix.O_NOCTTY|unix.O_NOFOLLOW|unix.O_CLOEXEC, 0666)
```

Comment:
Interesting. I feel like there's something relevant about SIGHUP that I probably don't know about. What's the idea behind the open/close?

Reply:
It's saying send SIGTERM instead of SIGHUP if `/dev/tty` can't be opened (there's no controlling terminal). I copied this from `lxc exec`, assuming they had run into cases where it was necessary. Here's the relevant lxd code change: https://github.com/lxc/lxd/commit/0f39ee55f525f7359a1bde32f1c519fbcfd0e883 ... I don't really understand when this would happen though, even after reading that commit message.

Reply:
Okay, so neither you nor me understand it. Let's please dig into it to make sure that's something we want to adopt.

Reply:
Sure. I've removed this special case for now, and asked Christian Brauner (author of that change) for more details on when this can happen / why it's necessary. I can always revert the commit to put it back in once we understand it and decide it's needed.

Reply:
Okay, I asked Christian and he said:

> `lxc exec` can cause the associated lxc process to hang if the controlling terminal is closed. This can happen when the command that is being run is ignoring `SIGHUP` signals sent when the controlling terminal was closed. This is the case e.g. when bash is running another command. This detects that there is no controlling terminal and sends SIGTERM instead of SIGHUP.

With the special handling for SIGHUP removed from Pebble, I was able to reproduce it by doing `pebble exec` on the following program that ignores SIGHUP (`gosleep.go` with imports elided):

```go
func main() {
    n, _ := strconv.Atoi(os.Args[1])
    fmt.Printf("sleeping %d seconds\n", n)
    signal.Ignore(syscall.SIGHUP)
    time.Sleep(time.Duration(n) * time.Second)
}
```

When I do `pebble exec -- gosleep 60` and then close the terminal, gosleep (and pebble exec) keep running. The special handling for SIGHUP that `lxc exec` fixes this -- it would send the process SIGTERM in this case, causing it to exit (I've tested that with the above program, and it works). The program exiting when the controlling terminal is closed seems like desired behavior to me, so I'd be in favour of reinstating this. Thoughts?

Reply:
Okay, Harry and I thought through this further, and we believe it's better to *not* kill the process, because it's explicitly ignoring SIGHUP. Instead, we forward SIGHUP (and the program may or may not ignore it) and then exit `pebble exec`. This is what ssh does, and seems reasonable.